### PR TITLE
Teams first iteration

### PR DESCRIPTION
**Pull request by aleksandr-wemakesoftware**
create global teams
join / leave a team
team search screen (mobile app)
team profile (mobile and web)

### CODE REVIEW

File: mobile-app/src/screens/Profile/Profile.js [hunk 1]
```diff
@@ -2,30 +2,50 @@ import React, { Component } from 'react';
...
+import SelectBlock from './SelectBlock';
 
 class Profile extends Component {
+
+  componentWillMount() {
+    this.props.getProfile();
+  }
+
+  handleTeamPress = team => {
+    this.props.navigation.navigate('TeamProfile', { team });
```
Comment (KucherenkoIhor) [inline mobile-app/src/screens/Profile/Profile.js]: hardcode

File: mobile-app/src/screens/Profile/Profile.js [hunk 2]
```diff
@@ -2,30 +2,50 @@ import React, { Component } from 'react';
...
+  };
+
+  handleJoinTeamPress = () => {
+    this.props.navigation.navigate('Teams');
```
Comment (KucherenkoIhor) [inline mobile-app/src/screens/Profile/Profile.js]: hardcode

File: mobile-app/src/screens/Profile/styles.js [hunk 3]
```diff
@@ -42,6 +42,65 @@ export default EStyleSheet.create({
...
+    flexWrap: 'wrap',
+  },
+  teamIconContainer: {
+    width: '20%',
+    flexGrow: 1,
+  },
+  teamIconImage: {
+    width: getWidthPercentage(40),
+    height: getWidthPercentage(40),
+    // borderRadius: getWidthPercentage(20),
```
Comment (KucherenkoIhor) [inline mobile-app/src/screens/Profile/styles.js]: remove commented code

File: mobile-app/src/screens/TeamProfile/TeamProfile.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  constructor(props) {
+    super(props);
+    const { params } = props.navigation.state;
+    this.state = params;
+  }
+
+  alert = (join, onPress, team) => {
+    Alert.alert(
+      `${join ? 'Join' : 'Leave'} a team`,
+      `Do you really want to ${join ? 'join' : 'leave'} the team ${team.name}?`,
```
Comment (KucherenkoIhor) [inline mobile-app/src/screens/TeamProfile/TeamProfile.js]: translation

File: mobile-app/src/screens/TeamProfile/TeamProfile.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+      [
+        { text: 'No', style: 'cancel' },
```
Comment (KucherenkoIhor) [inline mobile-app/src/screens/TeamProfile/TeamProfile.js]: translation

File: mobile-app/src/screens/Teams/Teams.js [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+  handleTeamPress = team => {
+    this.props.navigation.navigate('TeamProfile', { team });
+  };
+
+  render() {
+    const { t, teams, loading } = this.props;
+    const { search } = this.state;
+    return (
+      <View style={{ flex: 1 }}>
```
Comment (KucherenkoIhor) [inline mobile-app/src/screens/Teams/Teams.js]: extract style

File: mobile-app/src/screens/Teams/Teams.js [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+  componentWillMount() {
+    this.props.clearTeams();
+  }
+
+  getFilteredTeams = async () => {
+    const { search } = this.state;
+    const { teams, fetchTeams } = this.props;
+
+    if (search && search.length > 2) {
```
Comment (KucherenkoIhor) [inline mobile-app/src/screens/Teams/Teams.js]: what is 2?? It’s magic number

File: mobile-app/src/screens/Teams/Teams.js [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+      this.setState({ teams });
+    }
+  };
+
+  handleSearchChanged = (search) => {
+    this.setState({ search });
+  };
+
+  handleTeamPress = team => {
+    this.props.navigation.navigate('TeamProfile', { team });
```
Comment (KucherenkoIhor) [inline mobile-app/src/screens/Teams/Teams.js]: hardcode

File: mobile-app/src/screens/Teams/Teams.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            autoCorrect={false}
+            underlineColorAndroid={'#fff'}
+          />
+          <TouchableOpacity style={styles.searchButton}
+                            onPress={this.getFilteredTeams}>
+            <SimpleLineIcons name={'magnifier'} size={16} color={'#3E8EDE'}/>
+          </TouchableOpacity>
+        </View>
+        <View style={styles.container}>
+          {loading
```
Comment (KucherenkoIhor) [inline mobile-app/src/screens/Teams/Teams.js]: this condition is really big/ Extract to separate function and use if-else

File: mobile-app/src/screens/Teams/Teams.js [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+                      <Divider/>
+                    </View>
+                  ))}
+                </ScrollView> :
+                <View style={styles.defaultContainer}>
+                  <Text
+                    style={styles.defaultTextHeader}>{t('label_text_search_for_a_team')}</Text>
+                  <Text
+                    style={styles.defaultTextDescription}>{t('label_text_work_together')}</Text>
+                </View>}
```
Comment (KucherenkoIhor) [inline mobile-app/src/screens/Teams/Teams.js]: end of expression

File: web-app/src/pages/Home.js [hunk 11]
```diff
@@ -7,27 +7,35 @@ import {
...
 import userIcon from '../assets/user_menu.png';
+import teamsIcon from '../assets/teams_menu.png';
 import { Details } from '../components/Details';
 import { EditTrashpoint } from '../components/EditTrashpoint';
 import './Home.css';
 
 const SIDEBAR_LINKS = [
   { image: trashpointIcon, title: 'Trashpoints', selected: true },
   { image: userIcon, title: 'Users', selected: false },
+  { image: teamsIcon, title: 'Teams', selected: false },
```
Comment (KucherenkoIhor) [inline web-app/src/pages/Home.js]: hardcode

File: web-app/src/components/List/List.css [hunk 12]
```diff
@@ -4,6 +4,16 @@
     display: flex;
     flex-direction: column;
     overflow-y: auto!important;
```
Comment (vsheremet92) [inline web-app/src/components/List/List.css]: it's better not to use !important

File: web-app/src/pages/TeamDetails/TeamDetails.js [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+      return null;
+    }
+
+    return (
+      <div className="StatusCounts-container">
+        {_.map(STATUS_COUNT_HASH, (status, key) => {
+          const count = statusCounts[key] || 0;
+          return (
+            <div className="StatusCount-container" key={key}>
+              {console.log({ status, count })}
```
Comment (vsheremet92) [inline web-app/src/pages/TeamDetails/TeamDetails.js]: remove logs

File: web-app/src/pages/TeamDetails/TeamDetails.js [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+  };
+
+  oneOrMany = (number, str) => (number === 0 || number) && (number === 1 ? `${number} ${str}` : `${number} ${str}s`);
+
+  render() {
+
+    const { team, match: { params } } = this.props;
+
+    return team.id === params.id ?
+      <div className="AreaList">
```
Comment (vsheremet92) [inline web-app/src/pages/TeamDetails/TeamDetails.js]: different quotes for classnames. let's use single quotes or double quotes for every tag

File: web-app/src/pages/TeamDetails/TeamDetails.js [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+            <div className="TeamDetails-members">
+              <img className="TeamDetails-members-icon" src={imageMembers}/>
+              <p className='TeamDetails-members-p'>
+                {this.oneOrMany(team.members, 'member')}
+              </p>
+            </div>
+          </div>
+        </div>
+        <p className='TeamDetails-description-p'>{team.teamDescription}</p>
+        <div style={{ display: 'flex', flex: '1', flexDirection: 'column' }}>
```
Comment (vsheremet92) [inline web-app/src/pages/TeamDetails/TeamDetails.js]: it's better to keep styles in one format - css files or inline

File: web-app/src/pages/TeamsList/TeamsList.css [hunk 16]
```diff
@@ -0,0 +1,17 @@
+.TeamsList-sort-container {
+  padding: 10px;
+}
+
+.TeamsList-sort-button {
+  cursor: pointer;
+  margin-left: 5px;
+  color: #2672bd;
```
Comment (vsheremet92) [inline web-app/src/pages/TeamsList/TeamsList.css]: can we store color values in variables?

File: web-app/src/pages/TeamsList/TeamsList.js [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+
+  handleSearchChanged = event => {
+    this.setState({
+      search: event.target.value
+    });
+  };
+
+  handleSetSort = sortBy => {
+    this.setState({
+      sortBy: sortBy
```
Comment (vsheremet92) [inline web-app/src/pages/TeamsList/TeamsList.js]: you can use object property shorthand here

File: web-app/src/pages/TeamsList/TeamsList.js [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+    });
+  };
+
+  getFilteredTeams = () => {
+    const { search } = this.state;
+    const { teams } = this.props;
+    if (!search || (search && search.length < 3)) {
```
Comment (vsheremet92) [inline web-app/src/pages/TeamsList/TeamsList.js]: what is 3? maybe it's better to store this value in variable?

File: web-app/src/pages/TeamsList/TeamsList.js [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+          type="text"
+          value={this.state.search}
+          onChange={this.handleSearchChanged}
+          name="search"
+          placeholder="Team name"
+        />
+        <div className="TeamsList-sort-container">
+          Sort by:
+          <button
+            className={
```
Comment (vsheremet92) [inline web-app/src/pages/TeamsList/TeamsList.js]: for conditional classnames i'd recommend using https://www.npmjs.com/package/classnames
to improve code readability

File: web-app/src/pages/TeamsList/TeamsList.js [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+        </div>
+      </div>
+    );
+  }
+
+  render() {
+    const { search } = this.state;
+    return (
+      <List
+
```
Comment (vsheremet92) [inline web-app/src/pages/TeamsList/TeamsList.js]: unnecessary space

File: web-app/src/pages/TeamsList/TeamsList.js [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+const mapState = state => ({
+  teams: selectors.getAllTeams(state),
+  loading: selectors.getTeamsLoading(state),
+});
+const mapDispatch = {
+  fetchAllTeams: actions.fetchAllTeams,
+};
+
+TeamsList.propTypes = {
```
Comment (vsheremet92) [inline web-app/src/pages/TeamsList/TeamsList.js]: usage of propTypes is good practice, but why don't you use it for every component?

File: web-app/src/pages/TeamsList/components/TeamsListItem/TeamsListItem.css [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+  align-items: stretch;
+  justify-content: space-between;
+  padding-top: 13px;
+  padding-left: 8px;
+  padding-bottom: 9px;
+  cursor: default;
+}
+
+.TeamsListItem:hover {
+  background-color: #D8EDFF!important;
```
Comment (vsheremet92) [inline web-app/src/pages/TeamsList/components/TeamsListItem/TeamsListItem.css]: !important

File: web-app/src/reducers/trashpile/actions.js [hunk 23]
```diff
@@ -365,7 +365,6 @@ export const handleUpload = async ({ photos, markerId }) => {
 
     const handledPhotos = [...thumbnailsPhotos, ...mediumPhotos];
     const uploadedPhotosResponses = await uploadPhotosOnAzure(handledPhotos);
-
     if (uploadedPhotosResponses) {
       uploadedPhotosResponses.forEach(({ status }, index) => {
         const state = status === 201 ? 'confirmed' : 'failed';
```
Comment (vsheremet92) [inline web-app/src/reducers/trashpile/actions.js]: move status code to variable to make it clearer

File: web-app/src/reducers/user.js [hunk 24]
```diff
@@ -69,19 +69,20 @@ const authenticate = ({ network, token }) => async (dispatch, getState) => {
...
-        dispatch(appActions.showModal());
-      }
-    });
+    dispatch(fetchProfile())
+      .then(profile => {
+        if (!profile) {
+          return;
+        }
+        if (profile.locked) {
+          console.log('Account is locked');
```
Comment (vsheremet92) [inline web-app/src/reducers/user.js]: maybe we should use try/catch or throw for error handling?

File: web-app/src/routes/Home/Home.css [hunk 25]
```diff
@@ -62,11 +62,11 @@
 .Home-create-marker-button span {
     text-align: center;
     height: 20px;
-    min-width: 123px;
+    width: 123px;
     color: #FFFFFF;
```
Comment (vsheremet92) [inline web-app/src/routes/Home/Home.css]: you can use short notation here: #fff

File: web-app/src/routes/Home/Home.js [hunk 26]
```diff
@@ -59,11 +62,13 @@ class Home extends React.Component {
       <Switch>
         <Route path="/users/:id" exact component={UserDetails} />
         <Route path="/users" exact component={UserList} />
+        <Route path="/teams" exact component={TeamsList} />
         {/* <Route path="/areas/:id/trashpoints" exact render={TrashpointList} /> */}
```
Comment (vsheremet92) [inline web-app/src/routes/Home/Home.js]: remove comment

File: web-app/src/routes/Home/Home.js [hunk 27]
```diff
@@ -78,11 +83,13 @@ class Home extends React.Component {
 
   render() {
     const { userProfile } = this.props;
+    console.log({userProfile})
```
Comment (vsheremet92) [inline web-app/src/routes/Home/Home.js]: console

File: web-app/src/routes/Home/Home.css (hunk 25 above)
Comment (lynxlynxlynx) [reply web-app/src/routes/Home/Home.css]: More importantly, this was a bad merge resolution. The min-width change was introduced to fix a bug, in 14dc65a2e1b.

File: mobile-app/src/i18n/ui-strings.json [hunk 28]
```diff
@@ -127,7 +127,5 @@
   "label_text_about_4": "Estonian product, sponsored by the country and government",
   "label_text_about_5": "Funders: Estonian Ministry of Environment, The Ministry of Foreign Affairs of the Republic of Estonia, Estonian Republic 100 program",
   "label_button_editTP_delete": "Delete trashpoint",
-  "label_locked_account_warning": "Your account is locked. For details contact an administrator.",
-  "label_no_connection_note": "Without the Internet, you can only mark the trash. This mark will be transferred to the server when the connection to the Internet is restored.",
-  "label_your_coordinates": "Your coordinates"
+  "label_locked_account_warning": "Your account is locked. For details contact an administrator."
 }
```
Comment (lynxlynxlynx) [inline mobile-app/src/i18n/ui-strings.json]: bad merge resolution — the new strings should not be removed.

File: mobile-app/src/trans/en.json [hunk 29]
```diff
@@ -127,6 +134,21 @@
...
+  "label_text_search_teams": "Search teams",
+  "label_text_search_for_a_team": "Search for a team!",
+  "label_text_work_together": "Work together with other people and create competings teams. The winner is the world!",
+  "label_text_trashpoint": "Trashpoint",
   "label_no_connection_note": "Without the Internet, you can only mark the trash. This mark will be transferred to the server when the connection to the Internet is restored.",
   "label_your_coordinates": "Your coordinates"
-}
\ No newline at end of file
+
+}
```
Comment (lynxlynxlynx) [inline mobile-app/src/trans/en.json]: All these need to be added to mobile-app/src/i18n/ui-strings.json instead.

File: module-util/yarn-error.log [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+  moment@=2.18.1:
+    version "2.18.1"
+    resolved "https://registry.yarnpkg.com/moment/-/moment-2.18.1.tgz#c36193dd3ce1c2eed2adb7c802dbbc77a81b1c0f"
+  
+  uuid@=3.1.0:
+    version "3.1.0"
+    resolved "https://registry.yarnpkg.com/uuid/-/uuid-3.1.0.tgz#3dd3d3e790abc24d7b0d3a034ffababe28ebbc04"
+
+Trace: 
+  Error: ENOENT: no such file or directory, open '/module-util/node_modules/moment/CHANGELOG.md'
```
Comment (lynxlynxlynx) [inline module-util/yarn-error.log]: this file should be removed from the repo. No sense in tracking it.

File: backend-api/api/controllers/teams.js [hunk 31]
```diff
@@ -0,0 +1,28 @@
+'use strict';
+const senecaRequestMw = require('../../modules/swagger-setup/middleware/seneca');
+
+module.exports = {
+    getAllTeams: senecaRequestMw(
+        'role:db,cmd:getAllTeams',
+        req => ({
+            CC: req.swagger.params.country.value,
+            search: req.swagger.params.search && req.swagger.params.search.value,
```
Comment (artemarta) [inline backend-api/api/controllers/teams.js]: I'm not sure we need here this check. Search params are present always.

File: backend-db/plugins/teams.js [hunk 32]
```diff
@@ -0,0 +1,89 @@
...
+            .then(() => db.seedTeams(list))
+            .then(() => next())
+            .catch(e => next(e));
+    });
+
+    lucius.register('role:db,cmd:getAllTeams', async function (connector, args) {
+        return connector.input(args)
+            .use(async function ({CC, search}, responder) {
+                const teams = await db.getAllTeams();
+                const filteredTeams = search ? teams.filter(team =>
```
Comment (artemarta) [inline backend-db/plugins/teams.js]: I think we should move this filter into couch-db's map function.

File: backend-db/plugins/teams.js [hunk 33]
```diff
@@ -0,0 +1,89 @@
...
+                    team.name && team.name.toUpperCase().indexOf(search.toUpperCase()) > -1 ||
+                    team.teamDescription && team.teamDescription.toUpperCase().indexOf(search.toUpperCase()) > -1 ||
+                    COUNTRY_LIST[team.CC] && COUNTRY_LIST[team.CC].toUpperCase().indexOf(search.toUpperCase()) > -1
+                ) : teams;
+                const sortedTeams = _.sortBy(filteredTeams, 'name');
+                const sortedTeamsByCountry = _.sortBy(sortedTeams, team => team.CC !== CC);
```
Comment (artemarta) [inline backend-db/plugins/teams.js]: We can do sort also with couch-db.

File: backend-db/plugins/teams.js [hunk 34]
```diff
@@ -0,0 +1,89 @@
...
+                return responder.success(teamsWithMembers);
+            });
+    });
+
+    lucius.register('role:db,cmd:getAllTeamsWeb', async function (connector, args) {
+        return connector.input(args)
+            .use(async function ({CC, superadmin}, responder) {
+                const teams = await db.getAllTeams();
+                const filteredTeams = superadmin ? teams : teams.filter(team => team.CC === CC);
+                const sortedTeams = _.sortBy(filteredTeams, 'name');
```
Comment (artemarta) [inline backend-db/plugins/teams.js]: The same as above.

File: backend-db/plugins/teams.js [hunk 35]
```diff
@@ -0,0 +1,89 @@
...
+
+    lucius.register('role:db,cmd:getCountTeamsTrashpoints', async function (connector, args) {
+        return connector.input(args)
+            .use(async function ({}, responder) {
+                let teams = await db.getAllTeams();
+                teams.sort((a, b) => {
+                    const nameA = a.name.toUpperCase();
+                    const nameB = b.name.toUpperCase();
+                    return (nameA > nameB) ? 1 : (nameA < nameB ? -1 : 0)
+                });
```
Comment (artemarta) [inline backend-db/plugins/teams.js]: The same as above.

File: mobile-app/src/screens/Profile/Profile.js (hunk 1 above)
Comment (kakarda653) [reply mobile-app/src/screens/Profile/Profile.js]: @KucherenkoIhor 
Everywhere in application navigation happens like here.
Do you want me to create some const ROUTES and use it to navigate between screens like this?
`this.props.navigation.navigate(ROUTES.TEAM_PROFILE, { team });`

File: web-app/src/services/Api.js [hunk 36]
```diff
@@ -80,14 +80,15 @@ class ApiService {
...
-    : `${window.location.protocol}//api-${window.location.host}/api/v1`;
+export const BASE_URL = "http://server.noorsoft.ru:50000/api/v1"
+// export const BASE_URL =
+//   process.env.NODE_ENV === 'development'
+//     ? process.env.REACT_APP_LOCAL_API
+//       ? process.env.REACT_APP_LOCAL_API
+//       : 'https://api-qa.app.worldcleanupday.com/api/v1'
+//     : IS_PRODUCTION
+//     ? `${window.location.protocol}//api.${window.location.host}/api/v1`
+//     : `${window.location.protocol}//api-${window.location.host}/api/v1`;
```
Comment (lynxlynxlynx) [inline web-app/src/services/Api.js]: this shouldn't have been committed.

File: mobile-app/src/screens/Profile/Profile.js (hunk 1 above)
Comment (KucherenkoIhor) [reply mobile-app/src/screens/Profile/Profile.js]: Yap

**Comment by amangion** [comment]
Please resolve the conflicts

File: web-app/src/routes/Home/Home.css [hunk 37]
```diff
@@ -63,10 +63,11 @@
     text-align: center;
     height: 20px;
     min-width: 123px;
+    width: 123px;
```
Comment (lynxlynxlynx) [inline web-app/src/routes/Home/Home.css]: This looks like a bad conflict resolution. The width property was changed to min-width to fix an overflow issue. #168

**Pull request merged**